### PR TITLE
Issue#1433

### PR DESCRIPTION
**Pull request by Nathaniel-Hu**
This pull request has been updated to include the changes to the following documents:
- Requirements.hs (Haskell File)
- GamePhysics_SRS.tex (LaTeX File; stable)
- GamePhysics_SRS.html (HTML File; stable)

### CODE REVIEW

File: code/drasil-example/Drasil/GamePhysics/README.md [hunk 1]
```diff
@@ -32,9 +32,13 @@ README.md
   
 References.hs
   - References for the Game Physics example
-  
+
+Requirements.hs
+  - Requirements for the Game Physics example
+  - currently a work in progress (improvements/changes are being made to the NFRs)
```
Comment (smiths) [inline code/drasil-example/Drasil/GamePhysics/README.md]: Great to see Requirements.hs added to the repo.  Saying it is a work in progress is also a good idea.  However, I suggest removing the parenthetical part about NFRs.  That is a very current observation.  It won't age well.  Using the pull request and commit messages the fact that the NFRs have been updated will be captured in the project's history.

Comment (smiths) [reply code/drasil-example/Drasil/GamePhysics/README.md]: The date of the update of the README file should also be changed to reflect your edits.

File: code/drasil-example/Drasil/GamePhysics/Requirements.hs [hunk 2]
```diff
@@ -100,39 +100,36 @@ deterCollRespOverTime = cic "deterCollRespOverTime" deterCollRespOverTimeDesc "D
...
+
+performance :: ConceptInstance
+performance = cic "performance" (foldlSent [
+  S "The execution time for collision detection and collision resolution shall be", 
+  S "comparable to an existing 2D physics library on the market (e.g. Pymunk)"
+  ]) "Performance" nonFuncReqDom
+
+correctness :: ConceptInstance
+correctness = cic "correctness" (foldlSent [S "The", phrase output_ `sOf` S "simulation", 
+  S "results shall be compared to an existing implementation like Pymunk (please refer to: ", S "http://www.pymunk.org/en/latest/)"
```
Comment (smiths) [inline code/drasil-example/Drasil/GamePhysics/Requirements.hs:113]: Great!

File: code/drasil-example/Drasil/GamePhysics/README.md [hunk 3]
```diff
@@ -32,9 +32,12 @@ README.md
   
 References.hs
   - References for the Game Physics example
-  
+
+Requirements.hs
+  - Requirements for the Game Physics example
+
```
Comment (smiths) [inline code/drasil-example/Drasil/GamePhysics/README.md:38]: In a previous commit you mentioned that this is a work in progress.  That was a good detail to add.  You can add it back in.  :smile:

**Comment by JacquesCarette** [comment]
You know that you can do `make test` locally, right? This should have shown you that GamePhysics.Requirements needed some changes, before you checked it in.

**Comment by smiths** [comment]
@JacquesCarette has great feedback on using `make test`.  I just did a quick scan on the contributor's guide, and this isn't mentioned anywhere (as far as I can tell).  For issue #2136, please incorporate this information.

**Comment by Nathaniel-Hu** [comment]
@JacquesCarette, I ran `make test` as you suggested, and I got the following output: 

Screen Shot 2020-05-28 at 4 23 53 PM

Here it says that the generated outputs match the stable versions. Are there any specific errors I should be looking to address before I push these commits?

**Comment by JacquesCarette** [comment]
The build shows warnings in the compile, which are treated as errors. So while the code might work, it is not yet "clean". 

do
```
make clean
make test >& make.log
```
and then look at the warnings in the make.log file. (you can let the make output go to your screen too, but it does generated quite a lot, so it's easy to lose track of the warnings)

**Comment by smiths** [comment]
@Nathaniel-Hu, the advice above is also a good candidate for #2136.  :smile:

**Comment by Nathaniel-Hu** [comment]
Noted. Thank you so much!

**Comment by Nathaniel-Hu** [comment]
@JacquesCarette and @smiths,

I now have a version that works, and I will be uploading it soon to test it on Travis; however, a few unused imports had to be removed, as they were causing errors in the compilation stage. I will note this in my next commit.

**Comment by smiths** [comment]
Congratulations @Nathaniel-Hu on successfully passing the Travis build.  Now that you have a workflow, it will be easier in the future.  :smile:  I added #1433 as a linked issue to clarify what is motivating this PR.  That is something else that it would be good to make a habit (and add to the Contributor's Guide #2136).  :smile:

**Comment by Mornix** [comment]
Popping in here to add some more insight and clarify a few (small) things. 

Firstly, for anyone not aware, Travis logs can be viewed by either clicking the red 'x', yellow circle, or green checkmark, clicking view details, clicking the "This build" link on the page you're brought to. Travis runs every time code is pushed and anytime a pull request is opened. The Travis builds everything related to Drasil and the artefacts it produces. 

If we examine a completed Travis build, we'll notice yellow headings such as "CI: Build". These headings contain the terminal output (by clicking the arrow in the left margin) for a specific phase of building Drasil and its artefacts. We can also see that, for example, on this pull request, that @drasil-bot added the "CI: Build" label. This is done automatically from Travis builds to indicate build phases that encountered problems. 

One thing to note about Travis is it is more thorough and strict than local builds. Going through the phases, I'll explain what is done in each phase and where it may be stricter. 

### CI: HLint
Runs the HLint code linter against the Haskell source code. We do this to ensure the Drasil source code is idiomatic and consistent unifying individual contributors unique styles. This is not run locally as there may be intermediate Drasil changes that fail code style where the purpose was more functional and exploratory by a developer. 

### CI: Build
This phase is associated strictly with building the Drasil source code and not checking against stable. This part can be done locally with `make prog`, but it is generally advisable to run `make` (without any arguments) as it also checks stable which is a negligible time increase and is a very valuable check. A note about `make`, providing no arguments to `make` is an alias to manually running `make test`, but is five characters shorter (/cc @JacquesCarette). 

On Travis, the Drasil code is build in a way where warnings are elevated to errors (`-Werror` flag). This is an intentional difference. Locally, having warnings be errors can get in the way and be more of a hinderance than a useful check. You may be in the middle of a refactor and have some unused imports, the local build should still complete because the concern is "Is this refactor correct? Does it affect stable?" Once a commit is pushed to GitHub or a pull request is opened, the question should be less "Is this correct?" and more "Is this clean and concise?" While one warning isn't the end of the world, the more warnings that are introduced, the less a warning means and the less likely that a developer will notice the introduction of a warning if there's already a bunch. So really, the warnings-as-errors build on Travis is a way to prevent warnings from sneaking into the source. 

That being said, warnings are just that, every once in a blue moon a piece of code may be very intentional and desirable for some reason or extenuating circumstances. If you think that a warning is being emitted on a piece of code that is desirable, definitely talk to @JacquesCarette about it and he can advise you on the best course of action.

### CI: Stable
The last bit of when you run `make` locally. This checks that the `stable` folder matches the output of the `build` folder. If there are differences they will be printed in the Travis log. You can enable this printing differences to terminal in locally builds as `make NOISY=yes`.

### CI: TeX
Builds the Drasil generated LaTeX in the `build` folder. Drasil developers generally do not run this (`make tex`) all the time. My workflow was to run `make tex` before submitting a pull request to verify my changes visually in the generated LaTeX, or whenever I made changes to the LaTeX printer. 

### CI: Code
Builds the Drasil generated example code in the `build` folder. This is not mandatory and if you've never run this before you'll likely be spending a bunch of time installing various language compilers. My workflow use case for this target is similar to the LaTeX, where I would use this target after making changes to code gen to verify the generated code continues to compile. This target can be invoked with `make code`

### CI: Doxygen
This is another non-mandatory target. This target builds the code documentation produced in the generated code. Similar to TeX and Code targets, use this if you make changes to the generated documentation. This target can be invoked with `make doxygen`. 

### CI: Docs
This target builds the Haskell source documentation. It may be beneficial to run this target (`make docs`) locally every once in a while as documentation in Drasil changes, and to have a local copy. The documentation generated from this target can be found [on the Drasil website](https://jacquescarette.github.io/Drasil/docs/index.html) and should be at most a day behind the most recent commit to `master`.

### CI: Graphs
Target (`make graphs`) used almost exclusively for the [website](https://jacquescarette.github.io/Drasil/#packages) generation stuff. The graphs generated show the dependency graph of modules within a given Drasil sub-package.

### CI: Deploy
Generally empty on Travis builds unless a pull request touches the `website` directory. This target generated the Drasil website. You may want to run "most" of this target locally if you're modifying the website. Locally you'd want to run the `make deploy` which builds the website locally but does not perform the bits necessary to actually deploy the website which is designed to only be performed through Travis. 


Any of the above headings can appear on a pull request if that phase has failed during a pull request Travis build. The labels act as an indicator to @JacquesCarette about problems with the build and to the pull request author about where to look in the log or what targets to invoke locally. Further certain combinations of labels can indicate certain problems. For example, having just "CI: Build" (i.e. having no "CI: Stable" tag added) usually indicates the build portion failed with warnings. Another example is @smiths may want to defer reviewing a pull request with "CI: Stable" on it because that indicates that `stable` doesn't match the generated artefacts and there may be further changes to `stable`, which have yet to be made.

As an addendum, if you'd like to run HLint as part of a build, you can download and install HLint and run either as `hlint . && make`, which will prevent make from being run if HLint finds issues, or `hlint .; make` where the build will proceed regardless of the result of HLint. 

PS. @bmaclach has been actively developing and using Drasil more recently than I, so he may have some additional insight into workflows that work for him or new targets I may not be aware of.

**Comment by Nathaniel-Hu** [comment]
Thanks for the insight @Mornix! I'll be sure to look into it further for Issue #2136 (Updating the Contributor's Guide).

**Comment by smiths** [comment]
Yes, thank you @Mornix.  This is a great summary!  @Nathaniel-Hu is definitely correct that we want to incorporate this information into the Contributor's Guide!  @Nathaniel-Hu you should definitely reference this post by @Mornix.  You can link to it with:

https://github.com/JacquesCarette/Drasil/pull/2151#issuecomment-635750650

I'm thinking more and more that the Contributor's Guide should be written with an FAQ, or as a series of use cases.  There is so much information that new project members could get overwhelmed.

**Comment by JacquesCarette** [comment]
The required changes have been made, so I will merge it in.  Note that if this fixes 1433, that should be closed.

**Pull request merged**